### ISSUE TITLE
eksctl nodes have a bazillion private IPs

### ISSUE THREAD

**Issue by jwalton** [weaveworks/eksctl#1556]
**What help do you need?**

Sorry for the total n00b question, but I just launched an EKS cluster with three node groups, each with only one node, each in it's own AZ.  I'm launching into an existing VPC.  I happened to notice every node has ~20 private IPs attached to it:

![image](https://user-images.githubusercontent.com/1771003/68871627-bf3fa080-06ca-11ea-8693-2b8229c2058f.png)

What's going on here?  Is this expected behavior?

**Comment by res0nat0r** [comment]
Yes this is the AWS CNI plugin at work. It essentially pre-allocates IP's to your k8s nodes so that it can quickly spin up pods when they are assigned to the host. If you scale your podcount also up and down you'll see the number of attached secondary interfaces in your ENI list also change.

Some info here: https://github.com/aws/amazon-vpc-cni-k8s#eni-allocation

**Comment by jwalton** [comment]
Thanks for the explanation!  Was worried I'd done something horribly wrong.  :)

**Comment by res0nat0r** [comment]
No biggie 😄

**Comment by res0nat0r** [comment]
Yep, in fact that is a fundamental feature of k8s. I've found this blog post below very helpful, it also will give you a visual picture of what the AWS CNI plugin is doing to allow for AWS magic like VPC flow logs etc to work. Hope this helps.

https://www.contino.io/insights/kubernetes-is-hard-why-eks-makes-it-easier-for-network-and-security-architects

**Comment by res0nat0r** [comment]
Also relevant to ypur original question, from the link above:

> Virtual Machines in AWS have a max limit on the number of Elastic Network Interfaces that can be attached to any single vm at any time. This number varies by instance type. This can be calculated as follows:

> ENIs x Secondary ips per ENI = Max Pods per Node

> If the application you are building requires extremely high density of pods to nodes, you would be better off using one of the other CNI solutions like Calico or Weave, or perhaps something like Hashicorp’s Nomad which has been proven to scale up to one million containers across 5k hosts.

> Another issue with the AWS CNI is its inefficient use of IP addresses. In order to attach an IP address to a Pod as quickly as possible, the CNI pre-loads a number of IPs to the Node. If the cluster is deployed in a particularly small VPC, you may run out of addresses, even if they aren’t being used by real workloads.

**Comment by jwalton** [comment]
Thanks again so much for taking the time to answer my n00b question.  :)
